examples/step-69/step-69.cc

```diff
@@ -2734,57 +2781,101 @@ namespace Step69
+    // If the asynchronous writeback option is set we launch a background
+    // thread performing all the slow IO to disc. In that case we have to
+    // make sure that the background thread actually finished running. If
+    // not, we have to wait to for it to finish because we would otherwise
+    // overwrite <code>output_vector</code>, rerun the
+    // <code>schlieren_postprocessor</code> and
+    // DataOut<dim>::build_patches() prematurely before the output of the
+    // previous output cycle has been fully written out to disk.
+    //
+    // We launch said background thread with <a
+    // href="https://en.cppreference.com/w/cpp/thread/async"><code>std::async</code></a>
+    // that returns a <a
+    // href="https://en.cppreference.com/w/cpp/thread/future"><code>std::future</code></a>
+    // object. This <code>std::future</code> object contains the return
+    // value of the function, which is in our case simply
+    // <code>void</code>.
+
+    if (background_thread_state.valid())
       {
-        // We checkpoint the current state by doing the precise inverse
-        // operation to what we discussed for the <a href="Resume">resume
-        // logic</a>:
-
-        const unsigned int i =
-          discretization.triangulation.locally_owned_subdomain();
-        std::string name = base_name + "-checkpoint-" +
-                           Utilities::int_to_string(i, 4) + ".archive";
-
-        std::ofstream file(name, std::ios::binary | std::ios::trunc);
-
-        boost::archive::binary_oarchive oa(file);
-        oa << t << cycle;
-        for (const auto &it1 : U)
-          for (const auto &it2 : it1)
-            oa << it2;
+        TimerOutput::Scope timer(computing_timer, "main_loop - stalled output");
+        background_thread_state.wait();
       }
 
-    schlieren_postprocessor.compute_schlieren(U);
-
-    // The actual output code is standard. We create a (local) DataOut
-    // instance, attach all data vectors we want to output and finally
-    // call to DataOut<dim>::write_vtu_with_pvtu_record
-
-    DataOut<dim> data_out;
-    data_out.attach_dof_handler(offline_data.dof_handler);
-
     constexpr auto problem_dimension =
       ProblemDescription<dim>::problem_dimension;
-    const auto &component_names = ProblemDescription<dim>::component_names;
+
+    // At this point we make a copy of the state vector, run the schlieren
+    // postprocessor, and run DataOut<dim>::build_patches()
 
     for (unsigned int i = 0; i < problem_dimension; ++i)
-      data_out.add_data_vector(U[i], component_names[i]);
+      {
+        output_vector[i] = U[i];
+        output_vector[i].update_ghost_values();
+      }
 
-    data_out.add_data_vector(schlieren_postprocessor.schlieren,
-                             "schlieren_plot");
+    schlieren_postprocessor.compute_schlieren(output_vector);
 
     data_out.build_patches(discretization.mapping,
                            discretization.finite_element.degree - 1);
```

Comment:
Apparently this clears all previously computed patches. That, too, wasn't clear to me. 

My preference would have been if you create a `std::shared_ptr<DataOut>` at this place that you pass to the background thread. Once that thread is done, it automatically deletes the pointer. This also obviates the need for attaching vectors pro-actively.